src/app/fetch-status.service.ts

```diff
@@ -0,0 +1,122 @@
+import { Injectable, EventEmitter } from '@angular/core';
+import { HttpClient, HttpResponseBase } from '@angular/common/http';
+
+import { LinkData } from './link.service';
+import { BehaviorSubject } from 'rxjs';
+
+/**
+ * Internal interface to pair a status code with a boolean indicating success
+ * or failure
+ */
+interface Status {
+  code?: number;
+  ok?: boolean;
+}
+
+/**
+ * This service is responsible for checking the status of each link. For each
+ * unique link that is passed in in linkList it checks the HTTP status by sending
+ * a HEAD request. In addition in throttles the amount of requests that get sent
+ * to the HttpClient so it does not get overworked.
+ */
+@Injectable({
+  providedIn: 'root'
+})
+export class FetchStatusService {
+
+  private readonly batchSize = 50;
+  private activeRequests = 0;
```

Comment:
Is is possible to have a conflict if we call fetch() at the same time? Should we put this as a local variable similar to statusMap in fetch() ?

Reply:
We could I guess but then we might end up causing too much stress on the system. My thinking was that even if we have two concurrent fetch calls they should share an active requests variable since what we care about is that the total requests don't over stress the HttpClient